File: src/main/java/aurocosh/divinefavor/common/item/talismans/base/spell/SpellTalismanModPotionToggle.java
package aurocosh.divinefavor.common.item.talismans.base.spell;

import aurocosh.divinefavor.common.potions.base.effect.ModEffectToggle;
import aurocosh.divinefavor.common.potions.base.potion.ModPotionToggle;
import aurocosh.divinefavor.common.potions.base.potion.ModPotionToggleLimited;

public class SpellTalismanModPotionToggle extends ItemSpellTalisman {
    private final ModPotionToggle potion;

    public SpellTalismanModPotionToggle(String name, int startingSpellUses, ModPotionToggle potion) {
        super(name, startingSpellUses, true, true);
        this.potion = potion;
        if(potion instanceof ModPotionToggleLimited) {
            ModPotionToggleLimited limited = (ModPotionToggleLimited) potion;
            limited.setTalisman(this);
        }
    }

    @Override
    protected void performActionServer(TalismanContext context) {
        context.player.addPotionEffect(new ModEffectToggle(potion));
    }

    @Override
    public boolean isConsumeCharge(TalismanContext context) {
        return !context.player.isPotionActive(potion);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/item/spell_talismans/base/SpellTalismanModPotionToggle.kt
package aurocosh.divinefavor.common.item.spell_talismans.base

import aurocosh.divinefavor.common.item.spell_talismans.context.CastContext
import aurocosh.divinefavor.common.potions.base.effect.ModEffectToggle
import aurocosh.divinefavor.common.potions.base.potion.ModPotionToggle
import aurocosh.divinefavor.common.potions.base.potion.ModPotionToggleLimited
import aurocosh.divinefavor.common.spirit.base.ModSpirit

class SpellTalismanModPotionToggle(name: String, spirit: ModSpirit, favorCost: Int, private val potion: ModPotionToggle, val isCastFree: Boolean = false, val amplifier: Int = 0) : ItemSpellTalisman(name, spirit, favorCost, SpellOptions.ALL_CAST) {

    init {
        if (potion is ModPotionToggleLimited)
            potion.talisman = this
    }

    override fun performActionServer(context: CastContext) {
        val player = context.player
        if (player.isPotionActive(potion))
            player.removePotionEffect(potion)
        else
            player.addPotionEffect(ModEffectToggle(potion, amplifier))
    }

    public override fun isConsumeCharge(context: CastContext): Boolean {
        return !isCastFree && !context.player.isPotionActive(potion)
    }
}
